pkg/grpc/utils.go

```diff
@@ -35,67 +35,136 @@ var RPCTimeout = DefaultRPCTimeout
 
 var log = ctrl.Log.WithName("util")
 
-// CreateGrpcConnection create a grpc connection with given port and address
-func CreateGrpcConnection(address string, port int, caCertPath string, certPath string, keyPath string) (*grpc.ClientConn, error) {
-	if caCertPath != "" && certPath != "" && keyPath != "" {
-		var caCert, cert, key []byte
-		var err error
-		caCert, err = ioutil.ReadFile(caCertPath)
-		if err != nil {
-			return nil, err
-		}
-		cert, err = ioutil.ReadFile(certPath)
-		if err != nil {
-			return nil, err
-		}
-		key, err = ioutil.ReadFile(keyPath)
-		if err != nil {
-			return nil, err
-		}
-		return CreateGrpcConnectionFromRaw(address, port, caCert, cert, key)
+type GrpcBuilder struct {
+	options []grpc.DialOption
+	address string
+	port    int
+	err     error
+}
+
+func Builder() *GrpcBuilder {
+	return &GrpcBuilder{options: []grpc.DialOption{}, address: "localhost"}
+}
+
+func (it *GrpcBuilder) WithDefaultTimeout() *GrpcBuilder {
+	it.options = append(it.options, grpc.WithUnaryInterceptor(TimeoutClientInterceptor(DefaultRPCTimeout)))
+	return it
+}
+
+func (it *GrpcBuilder) WithTimeout(timeout time.Duration) *GrpcBuilder {
+	it.options = append(it.options, grpc.WithUnaryInterceptor(TimeoutClientInterceptor(timeout)))
+	return it
+}
+
+func (it *GrpcBuilder) Address(address string) *GrpcBuilder {
+	it.address = address
+	return it
+}
+
+func (it *GrpcBuilder) Port(port int) *GrpcBuilder {
+	it.port = port
+	return it
+}
+
+func (it *GrpcBuilder) Insecure() *GrpcBuilder {
+	it.options = append(it.options, grpc.WithInsecure())
+	return it
+}
+
+func (it *GrpcBuilder) TryTLSFromFiles(caCertPath string, certPath string, keyPath string) *GrpcBuilder {
```

Comment:
what is the difference between `TryTLSFromFiles` and `TLSFromFiles`?

Reply:
`TryTLSFromFiles` is somewhat similar to the original behavior of our implementation. If the specified path is not alright, it falls back to use an insecure connection. `TLSFromFiles` is more strict, it reports the error and refuse to continue rather than perform a fall back.